### DISCUSSION TITLE
google-cloud-sdk installs normally but requires "sudo" to uninstall

### DISCUSSION THREAD

**Discussion by jamesrtnz** [Homebrew/discussions#3900] [Casks]
I work in a corporate environment behind an SSL interception proxy, and this means (amongst other things) we need to use custom cacert file(s) with Google Cloud SDK; and that means creating an appropriate configuration BEFORE the "gcloud" command exists so that the "brew install google-cloud-sdk" command actually works correctly when it runs the Google-supplied installation script.

I was testing a shell script that will create an appropriate stub config file prior to the user running "brew install google-cloud-sdk"; and I testing by uninstalling/reinstalling the google-cloud-sdk cask.

However, on my macbook, the "brew uninstall google-cloud-sdk" command attempts to do the uninstall via SUDO; and I'm unsure why.  I don't normally have SUDO rights on my MacBook; and it doesn't install anything or run anything via SUDO when I install the google-cloud-sdk.

```
$ brew uninstall google-cloud-sdk
==> Uninstalling Cask google-cloud-sdk
==> Removing files:
/usr/local/Caskroom/google-cloud-sdk/latest/google-cloud-sdk
Password:
Sorry, user XXXXXX is not allowed to execute '/usr/bin/xargs -0 -- /bin/rm -r -f --' as root on HOSTNAME.
Error: Failure while executing; `/usr/bin/sudo -E -- /usr/bin/xargs -0 -- /bin/rm -r -f --` exited with 1. Here's the output:
Sorry, user XXXXXXX is not allowed to execute '/usr/bin/xargs -0 -- /bin/rm -r -f --' as root on HOSTNAME.
```

How can I find out what's causing the attempt to elevate via SUDO during the uninstall?

**Comment by mora-phi** [comment]
Same problem here, interested in a solution.

**Comment by carlocab** [comment]
This sounds related to Homebrew/brew#14593. It's not the same, but the underlying problem seems similar.

This might be worth filing an issue for at [Homebrew/brew](https://github.com/Homebrew/brew).

**Comment by jamesrtnz** [comment]
I have worked out why this happens.

The google-cloud-sdk.rb file contains the following code:

```
  # Not actually necessary, since it would be deleted anyway.
  # It is present to make clear an uninstall was not forgotten and that for this cask it is indeed this simple.
  uninstall delete: "#{staged_path}/#{token}"
```
This "uninstall delete:" request is resolved by the "def uninstall_delete(*paths, command: nil, **_)" defined in cask/artifact/abstract_uninstall.rb; and that has "sudo:true" hardcoded in it's call to the command.run function:

```
          command.run!(
            "/usr/bin/xargs",
            args:  ["-0", "--", "/bin/rm", "-r", "-f", "--"],
            input: resolved_paths.join("\0"),
            sudo:  true,
          )
```
hence, sudo is used.

I'm not enough of a ruby programmer to know how to pass using sudo as a flag into the "uninstall_delete" function, which would be my preferred fix.

**Comment by carlocab** [comment]
Try removing the `uninstall` line from the cask and see if uninstalling it still works:
```diff
diff --git a/Casks/google-cloud-sdk.rb b/Casks/google-cloud-sdk.rb
index dbb9922cd1..517a80b895 100644
--- a/Casks/google-cloud-sdk.rb
+++ b/Casks/google-cloud-sdk.rb
@@ -28,10 +28,6 @@ cask "google-cloud-sdk" do
   binary "#{token}/completion.zsh.inc",
          target: "#{HOMEBREW_PREFIX}/share/zsh/site-functions/_#{token}"
 
-  # Not actually necessary, since it would be deleted anyway.
-  # It is present to make clear an uninstall was not forgotten and that for this cask it is indeed this simple.
-  uninstall delete: "#{staged_path}/#{token}"
-
   caveats <<~EOS
     To add gcloud components to your PATH, add this to your profile:
```
You'll need to set `HOMEBREW_NO_INSTALL_FROM_API=1` for your changes to take effect.

**Comment by brettski** [comment]
I have also noticed that updating `google-cloud-sdk` via `brew upgrade` seems to uninstall, installed components. For example I needed to reinstall Cloud Pub/Sub Emulator after each update.
